src/cache/gcs.rs

```diff
@@ -96,19 +104,27 @@ impl Bucket {
         }))
     }
 
-    fn put(&self, key: &str, content: Vec<u8>, cred_provider: &GCSCredentialProvider) -> SFuture<()> {
+    fn put(&self, key: &str, content: Vec<u8>, cred_provider: &Option<GCSCredentialProvider>) -> SFuture<()> {
         let url = format!("{}/upload/storage/v1/b/{}/o?name={}&uploadType=media",
                     self.base_url,
                     percent_encode(self.name.as_bytes(), PATH_SEGMENT_ENCODE_SET),
                     percent_encode(key.as_bytes(), QUERY_ENCODE_SET));
 
         let client = self.client.clone();
 
-        Box::new(cred_provider.credentials(&client).and_then(move |creds| {
+        let creds_opt_future = if let &Some(ref cred_provider) = cred_provider {
+            future::Either::A(cred_provider.credentials(&self.client).map(Some))
+        } else {
+            future::Either::B(future::ok(None))
+        };
+
+        Box::new(creds_opt_future.and_then(move |creds_opt| {
             let mut request = Request::new(Method::Post, url.parse().unwrap());
             {
                 let mut headers = request.headers_mut();
-                headers.set(Authorization(Bearer { token: creds.token }));
+                if let Some(creds) = creds_opt {
+                    headers.set(Authorization(Bearer { token: creds.token }));
```

Comment:
I assume a put request isn't going to succeed without credentials, right? Should you make that explicitly fail here? I see that `GCSCache::put` will early-exit if it's configured in read-only mode.

One thing I've been meaning to add support for is better configuration of storage backends, so that you could have an explicitly read-only storage. I want this to support two-tier caching, where we can try to read from bucket A, and then fall back to reading from bucket B, and store new objects in bucket B.